First stab at a local scan

marumari opened the pull request:
Adds httpobs.scanner.local.scan().  This allows you to run synchronous scans without needing a database for storage, or redis/celery to conduct scans.

This addresses:
https://github.com/mozilla/http-observatory/issues/157
https://github.com/mozilla/http-observatory/issues/76
https://github.com/mozilla/http-observatory/issues/66

marumari commented:
Notes from the README:

## Running a scan from the local codebase, without DB, for continuous integration
```bash
# Install the HTTP Observatory
$ git clone https://github.com/mozilla/http-observatory.git
$ cd http-observatory
$ pip3 install .
$ pip3 install -r requirements.txt
```

```python
>>> from httpobs.scanner.local import scan
>>> scan('observatory.mozilla.org')  # a scan with default options
>>> scan('observatory.mozilla.org',  # all the custom options
         http_port=8080,
         https_port=8443,
         path='/foo/bar',
         cookies={'foo': 'bar'},
         headers={'X-Foo': 'bar'})
```

jarondl commented:
Looks great!
Bonus - should also solve #77.  I am just trying this out for servo/saltfs.

aneeshusa commented on README.md:
```diff
@@ -22,6 +22,26 @@ Sites can be scanned using:
+         http_port=8080,
+         https_port=8443,
+         path='/foo/bar',
+         cookies={'foo': 'bar'},
```
One suggestion would be to make it clear the `cookies` and `headers` args will be passed along with the request in the README, it wasn't clear to me until I looked at the docstring.

aneeshusa commented on httpobs/scanner/local.py:
```diff
@@ -0,0 +1,71 @@
+
+
+def scan(hostname, **kwargs):
+    # Always allow localhost scans when run in this way
```
Not sure but I think docstring extractors might break with the comment being first, maybe move it after the docstring?

marumari commented:
Did some code refactoring, and also added the ability to disable certificate verification; this will allow you to test HSTS/HPKP even on instances that might have self-signed certs that would get used during CI.

I also updated README.md to reflect @aneeshusa's comments.

marumari commented:
Added a small CLI to make CI and other things even easier:

```bash
$ PYTHONPATH='.' python3 httpobs/scanner/local.py
usage: local.py [-h] [--http-port HTTP_PORT] [--https-port HTTPS_PORT]
                [--path PATH] [--no-verify] [--cookies COOKIES]
                [--headers HEADERS] [--format FORMAT]
                hostname
local.py: error: the following arguments are required: hostname
```

```bash
$ PYTHONPATH='.' python3 httpobs/scanner/local.py google.com --format=report
Score: 35 [D]
Modifiers:
  Content Security Policy        [-25]  Content Security Policy (CSP) header not implemented
  Contribute                     [  0]  Contribute.json isn't required on websites that don't belong to Mozilla
  Cookies                        [-20]  Cookies set without using the Secure flag or set over http
  Cross Origin Resource Sharing  [  0]  Content is not visible via cross-origin resource sharing (CORS) files or headers
  Public Key Pinning             [ +5]  Preloaded via the HTTP Public Key Pinning (HPKP) preloading process
  Redirection                    [-10]  Redirects to https eventually, but initial redirection is to another http URL
  Strict Transport Security      [-10]  HTTP Strict Transport Security (HSTS) header set to less than six months (15768000)
  Subresource Integrity          [  0]  Subresource Integrity (SRI) not implemented, but all scripts are loaded from a similar origin
  X Content Type Options         [ -5]  X-Content-Type-Options header not implemented
  X Frame Options                [  0]  X-Frame-Options (XFO) header set to SAMEORIGIN or DENY
  X Xss Protection               [  0]  X-XSS-Protection header set to "1; mode=block"
```

```bash
$ PYTHONPATH='.' python3 httpobs/scanner/local.py google.com
{
    "scan": {
        "grade": "D",
        "likelihood_indicator": "MEDIUM",
        "response_headers": {
            "Alt-Svc": "quic=\":443\"; ma=2592000; v=\"36,35,34\"",
            "Cache-Control": "private, max-age=0",
            "Content-Encoding": "gzip",
            "Content-Type": "text/html; charset=UTF-8",
            "Date": "Thu, 03 Nov 2016 22:06:58 GMT",
            "Expires": "-1",
            "P3P": "CP=\"This is not a P3P policy! See https://www.google.com/support/accounts/answer/151657?hl=en for more info.\"",
            "Server": "gws",
            "Set-Cookie": "NID=90=XStpZ2fotK7ZzwP0lpC9QICbT6svRU-vPaOVojFbFQxMUlrOaXrpAb6KmYgX7ls4TZZNzMut5ojeTlTBCIoMiXjI6cDGQH1gM1HN0-ri1D4LCe7zc2T2Zb0oCViJc6hqpvbmiMWCAj9Jylc; expires=Fri, 05-May-2017 22:06:58 GMT; path=/; domain=.google.com; HttpOnly",
            "Strict-Transport-Security": "max-age=86400",
            "Transfer-Encoding": "chunked",
            "X-Frame-Options": "SAMEORIGIN",
            "X-XSS-Protection": "1; mode=block"
        },
        "score": 35,
        "tests_failed": 5,
        "tests_passed": 6,
        "tests_quantity": 11
    },
    "tests": {
        "content-security-policy": {
            "data": null,
            "expectation": "csp-implemented-with-no-unsafe",
            "pass": false,
            "result": "csp-not-implemented",
            "score_description": "Content Security Policy (CSP) header not implemented",
            "score_modifier": -25
        },
        "contribute": {
            "data": null,
            "expectation": "contribute-json-only-required-on-mozilla-properties",
            "pass": true,
            "result": "contribute-json-only-required-on-mozilla-properties",
            "score_description": "Contribute.json isn't required on websites that don't belong to Mozilla",
            "score_modifier": 0
        },
        "cookies": {
            "data": {
                "NID": {
                    "domain": ".google.com",
                    "expires": 1494022018,
                    "httponly": true,
                    "max-age": null,
                    "path": "/",
                    "port": null,
                    "secure": false
                }
            },
            "expectation": "cookies-secure-with-httponly-sessions",
            "pass": false,
            "result": "cookies-without-secure-flag",
            "score_description": "Cookies set without using the Secure flag or set over http",
            "score_modifier": -20
        },
        "cross-origin-resource-sharing": {
            "data": {
                "acao": null,
                "clientaccesspolicy": null,
                "crossdomain": null
            },
            "expectation": "cross-origin-resource-sharing-not-implemented",
            "pass": true,
            "result": "cross-origin-resource-sharing-not-implemented",
            "score_description": "Content is not visible via cross-origin resource sharing (CORS) files or headers",
            "score_modifier": 0
        },
        "public-key-pinning": {
            "data": null,
            "expectation": "hpkp-not-implemented",
            "includeSubDomains": true,
            "max-age": null,
            "numPins": null,
            "pass": true,
            "preloaded": true,
            "result": "hpkp-preloaded",
            "score_description": "Preloaded via the HTTP Public Key Pinning (HPKP) preloading process",
            "score_modifier": 5
        },
        "redirection": {
            "destination": "https://www.google.com/?gws_rd=ssl",
            "expectation": "redirection-to-https",
            "pass": false,
            "redirects": true,
            "result": "redirection-not-to-https-on-initial-redirection",
            "route": [
                "http://google.com/",
                "http://www.google.com/",
                "https://www.google.com/?gws_rd=ssl"
            ],
            "score_description": "Redirects to https eventually, but initial redirection is to another http URL",
            "score_modifier": -10,
            "status_code": 200
        },
        "strict-transport-security": {
            "data": "max-age=86400",
            "expectation": "hsts-implemented-max-age-at-least-six-months",
            "includeSubDomains": false,
            "max-age": 86400,
            "pass": false,
            "preload": false,
            "preloaded": false,
            "result": "hsts-implemented-max-age-less-than-six-months",
            "score_description": "HTTP Strict Transport Security (HSTS) header set to less than six months (15768000)",
            "score_modifier": -10
        },
        "subresource-integrity": {
            "data": {},
            "expectation": "sri-implemented-and-external-scripts-loaded-securely",
            "pass": true,
            "result": "sri-not-implemented-but-all-scripts-loaded-from-secure-origin",
            "score_description": "Subresource Integrity (SRI) not implemented, but all scripts are loaded from a similar origin",
            "score_modifier": 0
        },
        "x-content-type-options": {
            "data": null,
            "expectation": "x-content-type-options-nosniff",
            "pass": false,
            "result": "x-content-type-options-not-implemented",
            "score_description": "X-Content-Type-Options header not implemented",
            "score_modifier": -5
        },
        "x-frame-options": {
            "data": "SAMEORIGIN",
            "expectation": "x-frame-options-sameorigin-or-deny",
            "pass": true,
            "result": "x-frame-options-sameorigin-or-deny",
            "score_description": "X-Frame-Options (XFO) header set to SAMEORIGIN or DENY",
            "score_modifier": 0
        },
        "x-xss-protection": {
            "data": "1; mode=block",
            "expectation": "x-xss-protection-1-mode-block",
            "pass": true,
            "result": "x-xss-protection-enabled-mode-block",
            "score_description": "X-XSS-Protection header set to \"1; mode=block\"",
            "score_modifier": 0
        }
    }
}
```

I'll probably move it to a separate file tomorrow, but thought y'all might want to try it.  :)

marumari commented:
Okay, we're doooooooing it!  Please let me know if you change your requirements later and there are changes I can make to help.  :)
